Migrate schema tutorial files from GitLab MR !2286 and update mkdocs …

siamakn opened the pull request:
…config


Closes #56 
Closes #5

siamakn committed:
Migrate schema tutorial files from GitLab MR !2286 and update mkdocs config

hampusnasstrom commented on docs/tutorial/customization/custom_schemas.md:
```diff
@@ -0,0 +1,23 @@
+
+There are two primary approaches to creating custom schemas:
+
+- **Python schemas**: These offer highest possible level of customization by allowing the integration of **custom parsers** and **normalizers**, enabling maximum automation.
```
Again, not sure about "normalizers" here either. But also, I'm not sure you want to bring parsers into this.

hampusnasstrom commented on docs/tutorial/customization/custom_schemas.md:
```diff
@@ -0,0 +1,23 @@
+- [YAML schemas](custom_yaml.md): Defining structured YAML schemas for NOMAD.
+- [Custom ELN with YAML](custom_eln.md): Creating an ELN schema to enable structured metadata entry.
+- [Tabular parser in YAML](custom_tabular.md): Utilizing the built-in tabular parser to process spreadsheet data.
+- [Python schemas](custom_python.md): Exploring Python schemas and their advanced capabilities.
+- [Normalizers](custom_normalizer.md): Implementing normalizers to automate data processing.
```
What is the idea for these? Currently they are empty. Do you actually want to go into writing a normalizer? How will this tie in with the next tutorial, "developing a nomad plugin"?

siamakn replied:
These pages were part of the structure we had discussed internally in Area F, but since the overall structure is not finalized yet, I agree it makes sense to remove them for now. So I will remove them  both from the file and from nav in mkdocs.yaml. One can always reintroduce them once the tutorial structure and topics is confirmed.

hampusnasstrom commented on docs/tutorial/customization/custom_schemas.md:
```diff
@@ -0,0 +1,23 @@
+There are two primary approaches to creating custom schemas:
+
+- **Python schemas**: These offer highest possible level of customization by allowing the integration of **custom parsers** and **normalizers**, enabling maximum automation.
+- **Serialized schemas (YAML/JSON)**: Human-readable and easier to write and manage, ideal for defining customized ELNs without coding skills. However, they do not support normalizers.
```
I think a major thing for these is that they can run on the central NOMAD or on any Oasis without having to change the configuration of the Oasis. And also, "normalizer" again.

hampusnasstrom commented on docs/tutorial/customization/custom_yaml.md:
````diff
@@ -0,0 +1,164 @@
+      name: My NOMAD Custom Schema
+      description: This is a custom schema that includes several sections.
+    ```
+??? info "2. A schema can have several sections."
````
I wouldn't say that the `archive.yaml` file is *is* a schema. Rather the `archive.yaml` can contain a set of schema definitions. If you want to call it something it is technically a "schema package". For example inspecting this schema in NOMAD:
```yaml
definitions:
  name: Test
  sections:
    A:
      quantities:
        a:
          type: int
```
shows:
![image](https://github.com/user-attachments/assets/18dccd9f-5ae3-4889-aa3a-e160227019a1)

Note, that it says "package Test".

siamakn replied:
Thanks,  I didn't say "that the `archive.yaml`  file is a schema." 
The original sentence was "NOMAD YAML schema files must have the .archive.yaml extension for NOMAD to recognize them as schema files." My point was they should create a file with the extension `archive.yaml`. I made it clear, and included the suggested more precise term "schema package" and explanation.

hampusnasstrom commented on docs/tutorial/customization/custom_yaml.md:
````diff
@@ -0,0 +1,164 @@
+    ```
+
+
+??? info "3. Sections can inherit from NOMAD's `base_sections` or other sections."
````
This feels more advanced than quantities and subsections. I would have the order from simplest to most complex:
3. Quantities
4. Subsections
5. Inheritance

siamakn replied:
I can see the logic in ordering by complexity, but I thought it's still better to keep the current structure. Inheritance via base_sections is something all schemas rely on from the start (e.g., EntryData), and I think conceptually it is often easier to understand than the technicalities of subsections. This order was thought when we designed and introduced in Tutorial 16. I think it provides a more natural flow for first-time users (e.g., myself when I saw it for the first time).

hampusnasstrom commented on docs/tutorial/customization/custom_yaml.md:
```diff
@@ -0,0 +1,164 @@
+            - nomad.datamodel.metainfo.eln.Sample
+          quantities:
+            first_quantity:
+              - type: #For example, str or np.float64
+              - shape: #For example scalar or list (['*'])
+              - unit: #For example, meters, amperes, or seconds
```
Is there a reason you define this as a list of dictionaries rather than just directly a dict?
```suggestion
              type: #For example, str or np.float64
              shape: #For example scalar or list (['*'])
              unit: #For example, meters, amperes, or seconds
```

hampusnasstrom commented on docs/tutorial/customization/custom_yaml.md:
```diff
@@ -0,0 +1,164 @@
+      description: 
+
+      sections:
+        My_first_section:
+        My_second_section:
+        My_third_section:
```
These should be in PascalCase:
```suggestion
        MyFirstSection:
        MySecondSection:
        MyThirdSection:
```

hampusnasstrom commented on docs/tutorial/customization/custom_yaml.md:
```diff
@@ -0,0 +1,164 @@
+          quantities:
+
+          sub_sections:
+            My_first_subsection:
+              section:
+            My_second_subsection:
+              section:      
```
These should be in snake_case and should probably contain the actual section def as well:
```suggestion
            my_first_subsection:
              section: MySecondSection
            my_second_subsection:
              section: MyThirdSection
```

hampusnasstrom commented on docs/tutorial/customization/custom_yaml.md:
```diff
@@ -0,0 +1,164 @@
+          base_sections:
+            - nomad.datamodel.data.EntryData
+            - nomad.datamodel.metainfo.eln.Sample
+          quantities:
+            first_quantity:
+              - type: #For example, str or np.float64
+              - shape: #For example scalar or list (['*'])
+              - unit: #For example, meters, amperes, or seconds
+              m_annotations:
+                annotation_name:
+                  key1: value1  
+              sub_section:
+                My_first_subsection:
+                  section:
+                My_second_subsection:
+                  section:
+                My_third_subsection:
+                  section:
```
The indentation is wrong here.

hampusnasstrom commented on mkdocs.yml:
```diff
@@ -11,7 +11,13 @@ nav:
       - Exploring data: tutorial/explore.md
       - Access data via API: tutorial/access_api.md
-      - Schemas and plugins: tutorial/custom.md
+      - Schemas and plugins:
```
Not sure if this should really be called "Schemas and **plugins**" as we have another tutorial below on how to develop a plugin. What do you think?

hampusnasstrom commented on docs/tutorial/customization/custom_yaml.md:
```diff
@@ -0,0 +1,164 @@
+      sections:
+        My_first_section:
+          base_sections:
+            - nomad.datamodel.data.EntryData
+            - nomad.datamodel.metainfo.eln.Sample
```
We should definitely not use this section as an example as it is a deprecated section! I would recommend using only `ELNSample` as an example:
```suggestion
            - nomad.datamodel.metainfo.eln.ELNSample
```

hampusnasstrom reviewed (dismissed):
@siamakn I've gone through the first two pages for now and added my comments. I think we should discuss some things there before moving on.

hampusnasstrom commented on docs/tutorial/customization/custom_yaml.md:
```diff
@@ -0,0 +1,164 @@
+          base_sections:
+            - nomad.datamodel.data.EntryData
+            - nomad.datamodel.metainfo.eln.Sample
+          quantities:
+
```
I would add an example quantity already here with the one required argument, type.
```suggestion
          quantities:
            first_quantity:
              type: str # Uses python type notation like: int, np.float64, etc.
```

hampusnasstrom commented:
@ahm531 this is the PR from Siamak that should be merged in under the ELN tutorial somehow.

siamakn commented on docs/tutorial/customization/custom_schemas.md:
```diff
@@ -0,0 +1,23 @@
+- Define data structures that dictate how information is organized and presented.
+- Enable ELN functionality for manual data input and structured documentation.
+- Extract structured data from files using parsers, which allow NOMAD to read data from raw files.
+- Transform and enrich data post-upload using normalizers, ensuring compatibility with schema requirements and enhancing metadata consistency.
```
Thanks the comment and improving readability. I have clarified the sentence by linking directly to the relevant Glossary entries for parsers and normalizers. My intention was to use NOMAD's established terminology, as these tools are consistently referred to as normalizers throughout the documentation.

I agree that "normalization" could also work, but it might require explanation, which would add complexity to this short overview. However, I am happy to adjust it further if you think it is more precise and needed.

siamakn committed:
Addressed review comments

hampusnasstrom commented on docs/tutorial/customization/custom_schemas.md:
```diff
@@ -6,18 +6,17 @@ Custom schemas in NOMAD define how data is structured, stored, and displayed, al
 - Enable ELN functionality for manual data input and structured documentation.
 - Extract structured data from files using parsers, which allow NOMAD to read data from raw files.
-- Transform and enrich data post-upload using normalizers, ensuring compatibility with schema requirements and enhancing metadata consistency.
+- Transform and enrich data, post-upload, using [normalizers](../../reference/glossary.md#normalizer), ensuring compatibility with schema requirements and enhancing metadata consistency.
```
But do you really mean a "normalizer" here? Aren't you referring to the normalize method of a schema?

siamakn committed:
restructure and complete ELN tutorial and custom YAML schema tutorials

siamakn committed:
moved P3HT_optical.csv to directory under docs/tutoria/...

siamakn committed:
corrected the path to the csv file

siamakn committed:
Applying markdownlint fixes

siamakn committed:
fixed images path to pass the test. Fixed the test itself

siamakn commented:
Hi @yanghaoyu97@outlook.com @DanielYang59,

I modified the `tests/test_assets.py` while updating the ELN tutorial in the docs. **Locally** the test was telling ca. 81 images in `docs/tutorial/images` were “unused” even though they were referenced (e.g. pages under `docs/tutorial/eln/` linking via `../images/...`).

The newly generated pages pages live under `docs/tutorial/eln/`, but images live under `docs/tutorial/images/`. Those pages reference images via relative paths like `../images/...`.

The previous test only looked for images filenames  (assets in general)  in md files located in the *same directory level* as the `images/` or `data/` folder, so references from nested subfolders (in my case docs/tutorial/eln/*) could be missed (I hit this locally and it reported many images as "unused" even though they are referenced in the my md pages).

With this modification: for each `images/` and `data/` folder, it scans the md files in the **parent folder as well as its subfolders** (while skipping `images/` and `data/` during the scan).

Could you please confirm , or if you would prefer to keep the old behavior, I can revert this test changes.

DanielYang59 commented:
Hi @siamakn I think @blueraft implemented that in https://github.com/FAIRmat-NFDI/nomad-docs/pull/110 using Gemini CLI. But I guess it would be better if assets are scanned recursively?

ahm531 committed:
updated the tutorial_template

siamakn commented:
@DanielYang59 Thanks, Yes, it is exactly what I changed. 
@blueraft, I added you as reviewer. Could you please have a quick look at the modified `test_assets.py` this PR?

blueraft commented:
Could you create a new branch and copy over the `test_assets.py` changes? It would be easier to review without all of the other changes in this PR

ahm531 committed:
Reverted the changes to the order of sections & Move images to  an immediate sub-directory then updated links to images

ahm531 committed:
Revise tutorial content for creating a custom ELN schema (intro): enhance clarity, update learning objectives, and improve structure.

ahm531 committed:
Enhance tutorial on custom NOMAD ELN schema: add detailed example experiment, improve structure, and general revision

ahm531 committed:
Revise tutorial on creating a custom ELN schema: enhance clarity, update example experiment details, and improve structure.

ahm531 committed:
refinements

ahm531 committed:
reviewing the tabular_parser_yaml.md (general improvements, corrections, and making it consistent with the tutorials part of the docs

ahm531 committed:
Updated the tutorials admonitions legend with the checkpoint one

ahm531 committed:
updated admonition legend

ahm531 committed:
fixed the broken paths to images on the tutorials overview page

ahm531 committed:
removed link to 'Create custom schemas' tutorial from navigation

ahm531 committed:
remove overview sections from custom ELN schema and the tabular parser schema tutorials.

ahm531 committed:
update tutorial to remove optional designation from Step 7 on integrating the tabular parser section into an ELN template

ahm531 committed:
remove unnecessary comments and whitespace from tutorial markdown files (linting)

ahm531 committed:
linting

ahm531 committed:
linting again

ahm531 committed:
fixed the relocated images on the explore.md

ahm531 reviewed (changes requested):
@siamakn Thank you very much! This is great work. 
I have gone through the pages you have created and made several changes to improve consistency and clarity. 

**Main structural changes:** 

- Included the example admonitions for both in custom_eln_yaml.md and tabular_parser_yaml.md, to provide context of the scientific example used in the tutorials.
- “What you learn” come before “Before you Begin” - this was also corrected in the tutorials template. The Reasoning here is that the reader can first decide if it is something they are interested in reading, and then they evaluate if they can do it with the knowledge they have.
- Removed the section “A quick map of what your are building”. The path of the tutorial is provided concisely in the introductory part of each tutorial + the what you will learn. This paragraph is good and provides a more detailed and technical perspective of what will be done, however, I see the potential risk of introducing an overhead for the reader considering the technical terms not introduced. 
For this reason, I would stick the general introduction paragraph provided.
- Simplified the tasks when code is involved. The task in each step would to copy-paste a code snippet into a folder they are working with. This can be preceded with some introductory and concise statements, however, the “DO” task is to add the code to their schema file. 
After this, clarification for the various elements is provided as bullet list (as complete as needed - precise - and concise as possible).
- After each step, a checkpoint (admonition: success) is provided to show the current state of the file and to ensure that things are copied an pasted in the right place/indentation/order.

**Content changes:**
I have rewritten some parts to improve consistency with the other tutorials and provide more clarity while attempty stay concise. 

**Action requested:** 
The changes I made are in the recent commit to the PR. Please go through them once more, checking for typos, and any technical issues that I may have caused with me changes. 

Please update the imagesin the sliders. - Update them so that the font is Titillium web (consistent with the other images in the tutorial)
- This applies for the image slider in step7 (both in custom_eln_yaml.md and tabular_parser_yaml.md)

siamakn committed:
Remove unrelated changes from tests/test_assets.py

siamakn committed:
Polish ELN tutorial documentation, based on AM commit of Feb 27

siamakn committed:
fix md linting

siamakn committed:
Update ELN tutorial pages

siamakn commented:
Hi @ahm531 ,

 Thanks for the careful review and the structural improvements.
 
- I went through the recent changes again and checked the three tutorial pages for typos, wording consistency, and possible technical issues.
- I also verified that the YAML examples/indentation are consistent and that the docs build successfully with `mkdocs build --strict`, and in NOMAD.
- For the Step 7 slider images in `custom_eln_yaml.md` and `tabular_parser_yaml.md`, I checked the fonts in the svg file I have for all my images and they are all already using Titillium Web, so I didn't touch them.

ahm531 committed:
re-added the indentation after the list in before you begin

ahm531 reviewed (approved):
Thanks Siamak!
